core/quality/mysterium_morqa.go

```diff
@@ -18,45 +18,145 @@
-		qualityOracleAddress,
+		http:    &http.Client{Timeout: timeout},
+		baseURL: baseURL,
 	}
 }
 
 // ProposalsMetrics returns a list of proposals connection metrics
 func (m *MysteriumMORQA) ProposalsMetrics() []json.RawMessage {
-	req, err := requests.NewGetRequest(m.qualityOracleAddress, "proposals/quality", nil)
+	request, err := m.newRequestJSON(http.MethodGet, "proposals/quality", nil)
 	if err != nil {
 		log.Warn(mysteriumMorqaLogPrefix, "Failed to create proposals metrics request: ", err)
 		return nil
 	}
 
-	var metricsResponse ServiceMetricsResponse
-	err = m.http.DoRequestAndParseResponse(req, &metricsResponse)
+	response, err := m.http.Do(request)
 	if err != nil {
-		log.Warn(mysteriumMorqaLogPrefix, "Failed to request or parse proposals metrics: ", err)
+		log.Warn(mysteriumMorqaLogPrefix, err)
+		return nil
+	}
+	defer response.Body.Close()
+
+	var metricsResponse ServiceMetricsResponse
+	if err = parseResponseJSON(response, &metricsResponse); err != nil {
+		log.Warn(mysteriumMorqaLogPrefix, err)
 		return nil
 	}
 
 	return metricsResponse.Connects
 }
+
+// SendMetric submits new metric
+func (m *MysteriumMORQA) SendMetric(event *metrics.Event) error {
+	request, err := m.newRequestBinary(http.MethodPost, "metrics", event)
+	if err != nil {
+		return err
+	}
+
+	response, err := m.http.Do(request)
+	if err != nil {
+		return err
+	}
+	defer response.Body.Close()
+
+	if err = parseResponseError(response); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (m *MysteriumMORQA) newRequest(method, path string, body []byte) (*http.Request, error) {
+	url := m.baseURL
+	if len(path) > 0 {
+		url = fmt.Sprintf("%v/%v", url, path)
+	}
+
+	request, err := http.NewRequest(method, url, bytes.NewBuffer(body))
+	request.Header.Set("User-Agent", mysteriumMorqaAgentName)
+	request.Header.Set("Accept", "application/json")
+	return request, err
+}
+
+func (m *MysteriumMORQA) newRequestJSON(method, path string, payload interface{}) (*http.Request, error) {
+	payloadBody, err := json.Marshal(payload)
+	if err != nil {
+		return nil, err
+	}
+
+	req, err := m.newRequest(method, path, payloadBody)
+	req.Header.Set("Accept", "application/json")
+	return req, err
+}
+
+func (m *MysteriumMORQA) newRequestBinary(method, path string, payload proto.Message) (*http.Request, error) {
+	payloadBody, err := proto.Marshal(payload)
+	if err != nil {
+		return nil, err
+	}
+
+	request, err := m.newRequest(method, path, payloadBody)
+	request.Header.Set("Content-Type", "application/octet-stream")
+	return request, nil
+}
+
+func parseResponseJSON(response *http.Response, dto interface{}) error {
+	responseJSON, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return err
+	}
+	return json.Unmarshal(responseJSON, dto)
+}
+
+type errorDTO struct {
+	Message string `json:"message"`
+}
+
+// Sometimes we can get json message with single "message" field which represents error - try to get that
+func parseResponseError(response *http.Response) error {
```

Comment:
we just duplicating code here from http_client. Maybe make it more generic or merge with requests package and reuse here?

Reply:
https://github.com/mysteriumnetwork/node/pull/1160/files/463785708a6afc8344cd0940a41bc69434adf91b#r294191008

Reply:
Yes it has this function duplicated, but this API could have totally different error codes and DTOs also